File: core/mok_price.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
mok_price.py - MOKAGI 統一計費配置（唯一真源）

═══════════════════════════════════════════════════
⚠️  整個系統的收費標準都在此文件定義！
   修改此文件即可全局生效（後端計費 + 前端所有頁面顯示）。
═══════════════════════════════════════════════════

使用方式：
    後端 Python: from mok_price import price_per_token, cost_for_tokens, to_dict
    前端 JS:     mok_web.py 的 context_processor 自動注入
                 window.MOKAGI_PRICE（所有 Jinja2 模板可用）
                 或 fetch('/api/price') 動態獲取

計費規則：
    - 所有 AI 功能統一收費：HK$68 / 百萬 token
    - ai agent（人工智能助手）另收一次性安裝費：HK$5,000
    - 開發者（開源版）：完全免費
"""

# ========== 收費標準（唯一修改點）==========
MOKAGI_CURRENCY = "HKD"                 # 貨幣
MOKAGI_PRICE_HKD_PER_MILLION = 68       # 使用費：HK$68 / 百萬 token（所有 AI 功能統一收費）
MOKAGI_SETUP_FEE_HKD = 5000             # 一次性安裝費：HK$5,000（ai agent 版）
MOKAGI_GITHUB = "https://github.com/MOK2026/MOKAGI"

# ========== GPU（vast.ai 租機）計費 —— 侍女 GPU 用量帳單 ==========
# 真實成本 = 租機單價(US$/hr) × 租用時數（由 vast.ai API 取得）。
# 以下為「要給用戶看的費率」，尚未定案，先留空：
#   MOKAGI_GPU_MARKUP     成本倍率（例：2.0 = 收成本兩倍）; None = 未定（頁面顯示「費率待定」）
#   MOKAGI_GPU_USD_TO_HKD 美元兌港幣匯率，用於把成本換算成 HK$
MOKAGI_GPU_MARKUP = None
MOKAGI_GPU_USD_TO_HKD = 7.8  # 開源倉庫（開發者免費）
# =========================================


def price_per_token() -> float:
    """每 token 單價（HKD）"""
    return MOKAGI_PRICE_HKD_PER_MILLION / 1_000_000


def cost_for_tokens(tokens: float) -> float:
    """計算指定 token 數的費用（HKD）"""
    return tokens * price_per_token()


def display_per_million() -> str:
    """顯示用：如 'HK$68 / 百萬 token'"""
    return f"{MOKAGI_CURRENCY}${MOKAGI_PRICE_HKD_PER_MILLION} / 百萬 token"


def display_setup_fee() -> str:
    """顯示用：如 'HK$5,000'"""
    return f"{MOKAGI_CURRENCY}${MOKAGI_SETUP_FEE_HKD:,}"


def to_dict() -> dict:
    """返回完整價格信息（供 /api/price 與模板注入）"""
    return {
        "currency": MOKAGI_CURRENCY,
        "price_per_million": MOKAGI_PRICE_HKD_PER_MILLION,
        "price_per_token": price_per_token(),
        "setup_fee": MOKAGI_SETUP_FEE_HKD,
        "github": MOKAGI_GITHUB,
        "display": {
            "per_million": display_per_million(),
            "setup_fee": display_setup_fee(),
        },
        "gpu": {
            "markup": MOKAGI_GPU_MARKUP,
            "usd_to_hkd": MOKAGI_GPU_USD_TO_HKD,
            "display": ("成本 × %.2f（US$1=HK$%.2f）" % (MOKAGI_GPU_MARKUP, MOKAGI_GPU_USD_TO_HKD))
                       if MOKAGI_GPU_MARKUP else "費率待定",
        },
    }


if __name__ == "__main__":
    import json
    print(json.dumps(to_dict(), ensure_ascii=False, indent=2))

File: html/static/mok_price.js
/* ═══════════════════════════════════════════════════════════
 * mok_price.js - MOKAGI 統一計費（前端唯一讀取器）
 *
 * 唯一價格源：.mok/core/mok_price.py
 * 所有前端頁面透過本檔案讀取價格，修改 mok_price.py 即全局生效。
 *
 * 兩種讀取方式（自動偵測）：
 *   1. 模板注入：{{ MOKAGI_PRICE | tojson }} → window.MOKAGI_PRICE
 *   2. API 讀取：fetch('/api/price')（無注入時自動使用）
 *
 * 用法：
 *   await MokAgiPrice.ready()            // 等待價格就緒
 *   MokAgiPrice.perMillion               // 68（每百萬 token 價格）
 *   MokAgiPrice.perToken                 // 0.000068
 *   MokAgiPrice.setupFee                 // 5000
 *   MokAgiPrice.formatHKD(tokens)        // 計算並格式化費用
 *   MokAgiPrice.cost(tokens)             // 返回費用數值
 * ═══════════════════════════════════════════════════════════ */
window.MokAgiPrice = (function () {
    const _default = {
        currency: "HKD",
        price_per_million: 68,
        price_per_token: 0.000068,
        setup_fee: 5000,
        github: "https://github.com/MOK2026/MOKAGI",
        display: {
            per_million: "HK$68 / 百萬 token",
            setup_fee: "HK$5,000"
        }
    };

    let _price = null;          // 當前價格
    let _promise = null;        // 讀取 Promise（緩存）

    // ── 取得價格（優先模板注入，其次 API）──
    function load() {
        if (_price) return Promise.resolve(_price);
        if (window.MOKAGI_PRICE && window.MOKAGI_PRICE.price_per_million) {
            _price = window.MOKAGI_PRICE;
            return Promise.resolve(_price);
        }
        if (!_promise) {
            _promise = fetch('/api/price')
                .then(r => r.json())
                .then(data => { _price = data; return data; })
                .catch(() => { _price = _default; return _default; });
        }
        return _promise;
    }

    return {
        ready: load,
        get perMillion() { return _price ? _price.price_per_million : _default.price_per_million; },
        get perToken() { return _price ? _price.price_per_token : _default.price_per_token; },
        get setupFee() { return _price ? _price.setup_fee : _default.setup_fee; },
        get currency() { return (_price || _default).currency; },
        get displayPerMillion() { return (_price || _default).display.per_million; },
        get displaySetupFee() { return (_price || _default).display.setup_fee; },

        /** 計算 tokens 的費用（HKD 數值） */
        cost(tokens) {
            const p = _price ? _price.price_per_token : _default.price_per_token;
            return tokens * p;
        },

        /** 格式化費用為 HK$ 字串 */
        formatHKD(tokens) {
            const c = this.cost(tokens);
            if (c < 0.01) return 'HK$' + c.toFixed(4);
            if (c < 1) return 'HK$' + c.toFixed(3);
            return 'HK$' + c.toFixed(2);
        },

        /** 格式化費用（可選貨幣符號） */
        format(tokens) {
            return this.formatHKD(tokens);
        }
    };
})();
